상품 추가/조회/수정/삭제 기능

notmybussiness opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
1.Domain으로 Product 옮김
2. GlobalExceptionHandler에 CustomException들 이동 refactoring
3. BaseTimeEntity로 생성,수정시간 다룸
4. Product delete기능은 남김 -> 판매보류와 판매중 상태를 유지하는 STATUS를 변경하는 updateStatus 작성
5. 상품상세설명을 위한 Description 컬럼추가 및 기능 개발

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
제대로 merge되서 pr올라갔는지 모르겠네요..

notmybussiness committed:
test: Product Entity Red

notmybussiness committed:
test: Product Entity Green

notmybussiness committed:
test: Product entity Green all args test complete

notmybussiness committed:
test: Product Repository red complete & green ãin progress

notmybussiness committed:
Merge branch 'feature-product' of https://github.com/prgrms-be-devcourse/NBE3-4-1-Team02 into feature/product-api

Commit:
feat: 상품 엔티티 추가

Commit:
feat: 상품 엔티티 수정 및 패키지 구조 설정

notmybussiness committed:
test: ProductServiceTest , ProductRepositoryTest , ProductService&Repository complete

Commit:
feat: product 상품 목록 조회 기능 추가

Commit:
refactor:상품이 없을 경우 예외 처리 로직 추가

Commit:
chore: API 경로 설정 변경

Commit:
fix: h2 콘솔 설정 수정

Commit:
feat: 초기 데이터 DataInitializer 클래스 생성

Commit:
feat: 상품 목록 조회 기능 구현

Commit:
feat: 상품 목록 조회 기능 구현

Commit:
feat: product 상품 추가 기능 Controller, Service 추가

Commit:
feat: product 상품 추가 기능 Controller, Service 추가

Commit:
refactor: 상품 엔티티 및 DTO 변수명 변경

notmybussiness committed:
feat - 수정 삭제 기능 추가 테스트 x

notmybussiness committed:
test: - ì상품 수정,삭 기능 _ 없을늤트 통과

notmybussiness committed:
test- 상품 수정 삭제 ServiceTest 성공

notmybussiness committed:
refactor - Product Entity, ProductDTO í필 변경

notmybussiness committed:
refactor - product service 도 ce 필드명´서 변ê²변경

notmybussiness committed:
merge: 충돌 해결

notmybussiness committed:
merge: 충돌 해결

Commit:
feat:merge 후 변경사항 커밋

Commit:
fix : merge 후 service 코드 변경

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/product/service/ProductService.java:
```diff
@@ -0,0 +1,120 @@
+
+
+        if (products.isEmpty()) {
+            throw new ProductNotFoundException("No products found.");
```
특정 아이디가 아니라 말 그대로 모든 list 조회 시 비어있다면 NotFound예외를 던지기보다는 그냥 빈 리스트를 반환시켜 프론트에 넘겨주고 프론트에서 '상품이 없습니다' 같은 걸 화면에 보여주는 게 나을 것 같습니다.

seng-han replied:
그게 더 직관적인 방법 같네요 참고하겠습니다 !

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/product/service/ProductService.java:
```diff
@@ -0,0 +1,120 @@
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException("Product not found with id: " + id));
+
+        productRepository.delete(product);
```
데이터베이스에서 상품을 아예 지워버리게 될 경우

(당장은 없지만 만약 있다고 가정할 경우)
주문 내역을 조회할 때, 주문은 있지만 주문했던 상품이 안보이거나, 상품과 연계된 주문 데이터도 삭제되어 주문 기록도 사라지는 결과가 생길 거라 예상됩니다. 

Product 엔티티 클래스에 별도의 bollean 필드(isDeleted)를 만들어서 관리하는 게 좋다고 생각합니다.
-아마 status 필드가 그런 역할용으로 있는 것 같은데 맞나요?

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/product/controller/ProductController.java:
```diff
@@ -0,0 +1,47 @@
+    @GetMapping
+    public List<ProductDTO> findAllProducts(){
+        return productService.findAllProducts();
+    }
```
여기서만 ResponseEntity를 사용 안하시는 이유가 있으실까요?

seng-han replied:
상품 목록을 조회할 때 특별히 ResponseEntity를 사용하지 않고도 List<ProductDTO>만 반환하는 것도 충분하다고 생각했습니다.

InJunKangW reviewed (commented):
전체적으로 요구조건 잘 만족하는 코드라고 생각합니다.
몇몇 코멘트 남긴 부분들만 확인 부탁드려요

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/product/entity/Product.java:
```diff
@@ -0,0 +1,51 @@
+    @Column(nullable = false)
+    private boolean status;
+
+    @Column(nullable = false, updatable = false)
+    private LocalDateTime createdAt;
+
+    @Column(nullable = false)
+    private LocalDateTime updatedAt;
+
+    @PrePersist
+    protected void onCreate() {
+        this.createdAt = LocalDateTime.now();
+        this.updatedAt = LocalDateTime.now();
+    }
+
+    @PreUpdate
+    protected void onUpdate() {
+        this.updatedAt = LocalDateTime.now();
+    }
+}
```
아마 이 부분에 있어서 충돌이 발생할 것으로 예상이 됩니다.

저는 아래 방식으로 작성을 했었어요.

Spring Data JPA에서는 동일한 목적을 위해 @CreatedDate와 @LastModifiedDate 어노테이션을 사용하는 방식이 제공됩니다. 이를 활용하면 다음과 같은 이점을 얻을 수 있습니다:

- 중복 코드 제거: 별도의 메서드(onCreate, onUpdate) 없이 필드 값이 자동으로 설정됩니다.
- 코드 일관성: 여러 엔티티에서 동일한 BaseTimeEntity를 확장하여 일관된 방식으로 시간 관리를 처리할 수 있습니다.

```java
@Getter
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
public class BaseTimeEntity {

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(nullable = false)
    private LocalDateTime updatedAt;
}
```

```java
@Table(name = "product")
public class Product extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;
   ,
   ,
```

두 방식 모두 각각의 장단점이 있으니 기존 작성해주신 방식(@PrePersist, @PreUpdate)과 위 방식(BaseTimeEntity 활용) 중 한 가지를 선택하셔서 작성하시면 될 것 같습니다.

InJunKangW replied:
설명해주신 방식이 더 좋을 수 있다는 건 알겠는데
혹시 맨 처음의, "충돌이 발생할 것으로 예상" 이라는 게 어떤 의미인가요?

어플리케이션 내부 설정끼리의 충돌 (런타임 에러)을 의미하나요?
아니면 건희 님 코드와의 충돌을 의미하나요?

dnzp75 replied:
제가 pr올린 브랜치와의 충돌을 의미합니다

notmybussiness committed:
refactor: Product domain code strucure 개선
- Product entity 이동, baseTimeEntity 사용
- BaseTimeEntity ã상속 구조 적용
 - ãã package structure ãorganizing

notmybussiness committed:
refactor: Product 도메ìDomain code Structure 개선
-êException 통합

notmybussiness committed:
refactor: test코드들 정상화
- í 실제 api 작동 테스팅
- ë©api 작동 하지만, h2가 이전프로젝트 DB를 ã나타내는 문제

notmybussiness committed:
refactor: Product Entity 의 reduceStock 메소드 Service layer로 분리
- 그외에 yml 파일 프로파일 설정해서 개인화

Commit:
refactor:개별상품조회 메서드 제거

Commit:
feat: 상품상세조회 controller , service 구현

Commit:
refactor: 상품 목록 조회 시 없을 경우 빈 리스트 반환으로 수정

- 프론트엔드에서 빈 리스트를 처리할 수 있도록 변경

Commit:
feat: product description필드 추가

Commit:
feat:ProductDTO, ProductDescriptionDTO 나누어서 생성-m-description 필드를 포함한 ProductDescriptionDTO생성

Commit:
feat:ProductDTO, ProductDescriptionDTO 나누어서 생성

Commit:
Merge branch 'feature-product-1' of https://github.com/prgrms-be-devcourse/NBE3-4-1-Team02 into feature-product-1

Commit:
feat:DataInitializer에 Description 추가

Commit:
feat: 상품 상세 조회 시 description 필드 포함

notmybussiness committed:
feat: status ë변경기능 ì개발

notmybussiness committed:
test: status change code controllerTest fail red

notmybussiness committed:
test: Status change code ControllerTest & ServiceTest complete

notmybussiness committed:
stash

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/domain/product/service/ProductService.java:
```diff
@@ -16,16 +24,115 @@ public class ProductService {
+    }
+
+    //상품 목록 조회
+    private List<ProductDTO> convertToDTOList(List<Product> products) {
```
Page 반환도 있으면 좋을 것 같습니다.

notmybussiness committed:
merge: resolve conflicts by removing files

notmybussiness committed:
your commit message

notmybussiness committed:
feat: product branch merge
